### ISSUE TITLE
Implement tap method

### ISSUE THREAD

**Issue by jost125** [bonami/collections#48]
Tap is useful method to do sideeffect in the middle of chain for iterable collections (like logging). Eg:

```php
$ints
    ->map($power2)
    ->tap($log) // log in the middle
    ->map($multiply2)
```

It should have signature like

```php
/**
 * @param callable(T): void $sideEffect
 *
 * @return static<T>
 */
public function tap(callable $sideEffect): static
```

The semantic is, that tap **MUST NOT** mutate values inside of collection. It should either return the very same instance, or a new instance with same data (e.g. tap implemented on ArrayList can easily be implemented as each + return $this, and in that case, it can return the very same reference, but on LazyList, where the tap operation is deffered and does not materialize the collection, it should return a new reference of LazyList with contents unchanged.)

**Comment by jost125** [comment]
I have implemented tapEach for ArrayList, LazyList and Map. It also make sense to implement it on Option, TrySafe and Either (since they are all iterable).

It also might make sense to implement some specific tap operations:
- For Either: tapLeft (executes only if Either is left) and tapRight (executes only if Either is right AND should be just alias for tapEach, or we can ommit tapEach on Either and just have there the specific ones)
- For TrySafe tapSuccess and tapFailure (similar semantic like in Either).

Also, I have implemented tapEach on Map and I am not sure, whether it should tap whole pair. It might be better to have `tapEachKey`, `tapEachValue` and `tapEachPair`.
